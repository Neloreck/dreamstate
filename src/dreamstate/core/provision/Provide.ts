import { default as hoistNonReactStatics } from "hoist-non-react-statics";
import { ComponentType } from "react";

import { createManagersObserver } from "@/dreamstate/core/observing/createManagersObserver";
import { createClassWrapDecorator } from "@/dreamstate/polyfills/createClassWrapDecorator";
import { TDreamstateService } from "@/dreamstate/types";

/**
 * Decorator factory.
 * Provide context from context managers.
 * Observes changes and uses default react Providers for data flow.
 */
export function Provide<T extends ComponentType>(
  sources: Array<TDreamstateService<any>>
): ClassDecorator {
  return createClassWrapDecorator(function(targetClass: T): T {
    return hoistNonReactStatics(createManagersObserver(targetClass, sources), targetClass);
  });
}
