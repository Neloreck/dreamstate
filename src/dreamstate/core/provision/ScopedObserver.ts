import { createElement, ReactChild, ReactElement, useEffect, useMemo, useReducer } from "react";

import { CONTEXT_STATES_REGISTRY } from "@/dreamstate/core/internals";
import { startServiceObserving } from "@/dreamstate/core/observing/startServiceObserving";
import { stopServiceObserving } from "@/dreamstate/core/observing/stopServiceObserving";
import { forceUpdateReducer } from "@/dreamstate/core/provision/forceUpdateReducer";
import { addServiceObserverToRegistry } from "@/dreamstate/core/registry/addServiceObserverToRegistry";
import { registerService } from "@/dreamstate/core/registry/registerService";
import { removeServiceObserverFromRegistry } from "@/dreamstate/core/registry/removeServiceObserverFromRegistry";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

export interface IScopedObserverProps<T> {
  children?: ReactChild;
  initialState?: T;
  manager: TAnyContextManagerConstructor;
  dependencies: Array<TAnyContextManagerConstructor>;
}

/**
 * Provider that observes
 */
export function ScopedObserver<T>({
  children,
  initialState,
  dependencies,
  manager
}: IScopedObserverProps<T>): ReactElement {
  const [ , onUpdateNeeded ] = useReducer(forceUpdateReducer, null);

  /**
   * Use memo for first and single init of required components.
   * useLayoutEffect will not work for some environments and SSR.
   *
   * Note: Shared between components that do mount-unmount is the same node.
   */
  useMemo(function(): void {
    registerService(manager, initialState);
  }, [ dependencies ]);

  /**
   * Mount current observers.
   * Count references of providers to detect whether we start provisioning or ending it.
   *
   * todo: check option where we do not unload state on HMR reload
   * todo: check option where we trigger full reload on HMR with scoped providers
   */
  useEffect(function() {
    addServiceObserverToRegistry(manager, onUpdateNeeded);
    registerService(manager, initialState);
    startServiceObserving(manager);

    return function() {
      removeServiceObserverFromRegistry(manager, onUpdateNeeded);
      stopServiceObserving(manager);
    };
  }, [ dependencies ]);

  return createElement(manager.REACT_CONTEXT.Provider, { value: CONTEXT_STATES_REGISTRY.get(manager) }, children);
}
