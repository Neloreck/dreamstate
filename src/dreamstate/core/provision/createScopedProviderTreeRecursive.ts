import { createElement, ReactNode } from "react";

import { ScopedProvider } from "@/dreamstate/core/provision/ScopedProvider";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Subtree provider as global scope helper.
 * Recursive impl.
 */
export function createScopedProviderTreeRecursive<T extends TAnyObject>(
  bottom: ReactNode = null,
  sources: Array<TAnyContextManagerConstructor>,
  current: number,
  initialState?: T
): ReactNode {
  return current >= sources.length
    ? bottom
    : createElement(
      ScopedProvider,
      { manager: sources[current], initialState },
      createScopedProviderTreeRecursive(bottom, sources, current + 1, initialState)
    );
}
