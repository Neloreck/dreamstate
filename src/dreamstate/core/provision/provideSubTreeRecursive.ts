import { createElement, ReactElement } from "react";

import { CONTEXT_STATES_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Subtree provider as global scope helper.
 * Recursive impl.
 */
export function provideSubTreeRecursive(
  bottom: ReactElement,
  sources: Array<TAnyContextManagerConstructor>,
  current: number
): ReactElement {
  return current >= sources.length
    ? bottom
    : createElement(
      sources[current].REACT_CONTEXT.Provider,
      { value: CONTEXT_STATES_REGISTRY.get(sources[current]) },
      provideSubTreeRecursive(bottom, sources, current + 1)
    );
}
