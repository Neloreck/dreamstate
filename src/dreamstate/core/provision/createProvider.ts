import { FunctionComponent, ReactNode, useCallback, useState } from "react";

import { EMPTY_ARR } from "@/dreamstate/core/internals";
import { useHotObservers } from "@/dreamstate/core/observing/useHotObservers";
import { useStaticObservers } from "@/dreamstate/core/observing/useStaticObservers";
import { provideSubTreeRecursive } from "@/dreamstate/core/provision/provideSubTreeRecursive";
import { ContextManager } from "@/dreamstate/core/services/ContextManager";
import { ContextService } from "@/dreamstate/core/services/ContextService";
import {
  TAnyContextManagerConstructor,
  TAnyContextServiceConstructor,
  TAnyObject
} from "@/dreamstate/types";

export interface IProviderProps<T> {
  initialState?: T;
  partialHotReplacement?: boolean;
  children?: ReactNode;
}

/**
 * Create component for manual provision without HOC/Decorator-like api.
 * Useful if your root is functional component or you are using createComponent api without JSX.
 */
export function createProvider<
  T extends TAnyObject = TAnyObject
>(
  sources: Array<TAnyContextServiceConstructor>
): FunctionComponent<IProviderProps<T>> {
  if (!Array.isArray(sources)) {
    throw new TypeError(
      "Wrong providers parameter supplied. Only array of context services is acceptable."
    );
  }

  for (let it = 0; it < sources.length; it ++) {
    if (!sources[it] || !(sources[it].prototype instanceof ContextService)) {
      throw new TypeError("Only classes extending ContextService can be supplied for provision.");
    }
  }

  /**
   * Check only managers with required provision.
   * Do not include services for subTree rendering but add registering logic for services.
   */
  const managers: Array<TAnyContextManagerConstructor> = sources.filter(function(
    Service: TAnyContextServiceConstructor
  ) {
    return Service.prototype instanceof ContextManager;
  }) as Array<TAnyContextManagerConstructor>;

  /**
   * Create observer component that will handle observing.
   */
  function Observer(props: IProviderProps<T>): ReactNode {
    const [ , forceRender ] = useState({});
    const updateProviders = useCallback(function() {
      forceRender({});
    }, EMPTY_ARR);

    if (props.partialHotReplacement) {
      useHotObservers(sources, props.initialState, updateProviders);
    } else {
      useStaticObservers(sources, props.initialState, updateProviders);
    }

    return provideSubTreeRecursive(props.children, managers, 0);
  }

  if (IS_DEV) {
    Observer.displayName = `Dreamstate.Observer[${sources.map(function(it: TAnyContextServiceConstructor) {
      return it.name;
    })
    }]`;
  } else {
    Observer.displayName = "DS.Observer";
  }

  return Observer as FunctionComponent<IProviderProps<T>>;
}
