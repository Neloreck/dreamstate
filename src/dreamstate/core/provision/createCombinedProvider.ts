import { FunctionComponent, ReactNode } from "react";

import { IProviderProps } from "@/dreamstate/core/provision/createProvider";
import { provideSubTreeRecursive } from "@/dreamstate/core/provision/provideSubTreeRecursive";
import { useSourceObserving } from "@/dreamstate/core/provision/useSourceObserving";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Create provider that unifies all data sources and tries to re-render providers on every update.
 */
export function createCombinedProvider<T extends TAnyObject>(
  sources: Array<TAnyContextManagerConstructor>
): FunctionComponent<IProviderProps<T>> {
  function Observer(props: IProviderProps<T>): ReactNode {
    const registry: Map<TAnyContextManagerConstructor, TAnyObject> = useSourceObserving(sources, props);

    return provideSubTreeRecursive(props.children, sources, registry);
  }

  if (IS_DEV) {
    Observer.displayName = `Dreamstate.Observer[${sources.map(function(it: TAnyContextManagerConstructor) {
      return it.name;
    })
    }]`;
  } else {
    Observer.displayName = "DS.Observer";
  }

  return Observer as FunctionComponent<IProviderProps<T>>;
}
