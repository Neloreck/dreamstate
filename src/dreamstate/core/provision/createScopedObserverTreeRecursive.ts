import { createElement, ReactNode } from "react";

import { IProviderProps } from "@/dreamstate/core/provision/createProvider";
import { ScopedObserver } from "@/dreamstate/core/provision/ScopedObserver";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Subtree provider as global scope helper.
 * Recursive impl.
 */
export function createScopedObserverTreeRecursive<T extends TAnyObject>(
  sources: Array<TAnyContextManagerConstructor>,
  props: IProviderProps<T>,
  current: number = 0
): ReactNode {
  return current >= sources.length
    ? props.children
    : createElement(
      ScopedObserver,
      { manager: sources[current], initialState: props.initialState, dependencies: sources },
      createScopedObserverTreeRecursive(sources, props, current + 1)
    );
}
