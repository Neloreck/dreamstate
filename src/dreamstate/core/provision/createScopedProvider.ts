import { FunctionComponent, ReactNode } from "react";

import { IProviderProps } from "@/dreamstate/core/provision/createProvider";
import { createScopedProviderTreeRecursive } from "@/dreamstate/core/provision/createScopedProviderTreeRecursive";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

export function createScopedProvider<T extends IProviderProps<any>>(
  sources: Array<TAnyContextManagerConstructor>
): FunctionComponent<IProviderProps<T>> {
  function Observer(props: T): ReactNode {
    return createScopedProviderTreeRecursive(props.children, sources, 0, props.initialState);
  }

  Observer.displayName = "DS.Provider";

  return Observer as FunctionComponent<IProviderProps<T>>;
}
