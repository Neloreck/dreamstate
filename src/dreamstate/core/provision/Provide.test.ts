import { mount } from "enzyme";
import { createElement, PureComponent, ReactNode } from "react";

import { CONTEXT_SERVICES_ACTIVATED } from "@/dreamstate/core/internals";
import { Provide } from "@/dreamstate/core/provision/Provide";
import { TestContextManager, TestContextService, TestSingleContextService } from "@/fixtures";

describe("@Provide decorator.", () => {
  @Provide([ TestContextManager, TestContextService, TestSingleContextService ])
  class ExampleProvided extends PureComponent {

    public render(): ReactNode {
      return "test";
    }

  }

  it("Should render correct component tree.", async () => {
    expect(CONTEXT_SERVICES_ACTIVATED.has(TestContextService)).toBeFalsy();
    expect(CONTEXT_SERVICES_ACTIVATED.has(TestContextManager)).toBeFalsy();
    expect(CONTEXT_SERVICES_ACTIVATED.has(TestSingleContextService)).toBeFalsy();

    const tree = mount(createElement(ExampleProvided, {}));

    expect(tree).toMatchSnapshot();

    expect(CONTEXT_SERVICES_ACTIVATED.has(TestContextService)).toBeTruthy();
    expect(CONTEXT_SERVICES_ACTIVATED.has(TestContextManager)).toBeTruthy();
    expect(CONTEXT_SERVICES_ACTIVATED.has(TestSingleContextService)).toBeTruthy();

    tree.unmount();

    expect(CONTEXT_SERVICES_ACTIVATED.has(TestContextService)).toBeFalsy();
    expect(CONTEXT_SERVICES_ACTIVATED.has(TestContextManager)).toBeFalsy();
    expect(CONTEXT_SERVICES_ACTIVATED.has(TestSingleContextService)).toBeTruthy();
  });
});
