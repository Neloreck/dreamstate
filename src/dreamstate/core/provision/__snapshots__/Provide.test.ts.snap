// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Provide decorator Should render correct component tree 1`] = `
<Memo(DS.Observer)>
  <ExampleProvided>
    test
  </ExampleProvided>
</Memo(DS.Observer)>
`;
